File: 프로그래머스/삼각 달팽이/triangle.cpp
#include <string>
#include <vector>

using namespace std;

vector<int> solution(int n) {
    vector<int> answer;


    int loops = n / 3;
    int index = n % 3;
    vector<vector<int>> arry;
    for (int i = 1; i <= n; i++) {
        vector<int> temprow(i, 0);
        arry.push_back(temprow);
    }
    int curx = 0;
    int cury = 0;
    int curn = n;
    int curnum = 0;
    for (int i = 0; i < loops; i++) {
        for (int j = 0; j < curn - 1; j++) {
            curnum++;
            arry[cury++][curx] = curnum;
        }
        for (int j = 0; j < curn; j++) {
            curnum++;
            arry[cury][curx++] = curnum;
        }
        curx -= 2;
        cury--;
        for (int j = 0; j < curn - 2; j++) {
            curnum++;
            arry[cury--][curx--] = curnum;
        }
        cury += 2;
        curx++;
        curn -= 3;
    }
    if (index == 1) {
        curnum++;
        arry[cury][curx] = curnum;
    }
    else if (index == 2) {
        curnum++;
        arry[cury][curx] = curnum;
        arry[cury + 1][curx] = curnum + 1;
        arry[cury + 1][curx + 1] = curnum + 2;
    }
    for (int i = 0; i < arry.size(); i++) {
        for (int j = 0; j < arry[i].size(); j++) {
            answer.push_back(arry[i][j]);
        }
    }
    return answer;
}

File: 프로그래머스/삼각 달팽이/triangle.java
package com.company;
    class Solution {
        public int[] solution(int n) {
            int [][] arry = new int[n][n];
            int x=0;
            int y=-1;
            // 0(아래, 오른 대각)
            int flag = 0;
            int curnum =1;
            for(int i=n; i>=1;i--){
                for(int j=1;j<=i;j++){
                    if(flag%3==0){
                        y++;
                        arry[y][x] = curnum++;
                    }
                    else if(flag%3==1){
                        x++;
                        arry[y][x] = curnum++;
                    }
                    else{
                        x--;
                        y--;
                        arry[y][x] = curnum++;
                    }
                }
                flag++;
            }
            int [] answer = new int[--curnum];
            int idx =0;
            for(int i=0; i<n;i++){
                for(int j=0; j<=i; j++){
                    answer[idx++]=arry[i][j];
                }
            }
            return answer;
        }
    }
public class Main {
    public static void main(String[] args) {
        Solution a = new Solution();
        a.solution(4);

    }
}
